x/programs/examples/imports/program/program.go

```diff
@@ -0,0 +1,191 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package imports
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/bytecodealliance/wasmtime-go/v12"
+
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/state"
+	"github.com/ava-labs/hypersdk/x/programs/examples/storage"
+	"github.com/ava-labs/hypersdk/x/programs/runtime"
+)
+
+const Name = "program"
+
+type Import struct {
+	db         state.Immutable
+	log        logging.Logger
+	imports    runtime.Imports
+	meter      runtime.Meter
+	registered bool
+}
+
+// New returns a new program invoke host module which can perform program to program calls.
+func New(log logging.Logger, db state.Immutable) *Import {
+	return &Import{
+		db:  db,
+		log: log,
+	}
+}
+
+func (i *Import) Register(linker runtime.Link, meter runtime.Meter, imports runtime.Imports) error {
+	if i.registered {
+		return fmt.Errorf("import module already registered")
+	}
+	if err := linker.FuncWrap(Name, "invoke_program", i.invokeProgramFn); err != nil {
+		return err
+	}
+	i.registered = true
+	i.imports = imports
+	i.meter = meter
+
+	return nil
+}
+
+// invokeProgramFn makes a call to an entry function of a program in the context of another program's ID.
+func (i *Import) invokeProgramFn(
+	caller *wasmtime.Caller,
+	programIDPtr,
+	functionPtr,
+	functionLen,
+	argsPtr,
+	argsLen uint32,
+	maxUnits uint64,
+) int64 {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	memory := runtime.NewMemory(runtime.NewExportClient(caller))
+
+	programIDBytes, err := memory.Range(uint32(programIDPtr), uint32(ids.IDLen))
+	if err != nil {
+		i.log.Error("failed to read function name from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	programID, err := ids.ToID(programIDBytes)
+	if err != nil {
+		i.log.Error("failed to create programID",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	// get the entry function for invoke to call.
+	functionBytes, err := memory.Range(uint32(functionPtr), uint32(functionLen))
+	if err != nil {
+		i.log.Error("failed to read function name from memory",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	programWasmBytes, exists, err := storage.GetProgram(i.db, programID)
+	if !exists {
+		i.log.Error("program does not exist")
+	}
+	if err != nil {
+		i.log.Error("failed to get program from storage",
+			zap.Error(err),
+		)
+		return -1
+	}
+
+	// spend the maximum number of units allowed for this call
```

Comment:
nit: can you elaborate more in this comment on how you do this (spend max, run, refund difference)?

Reply:
And that this is required because we have different meters?

Reply:
I extended the meter interface and added a `TransferUnits` method along with docs. TL;DR yes each runtime has an internal meter. Each level we go deeper that accounting must be kept. I think this is fine because all of this logic will be in `programs`.

Reply:
But we do need a rollback on failure mechanism to ensure unspent units are returned to parent on failure. Will add a TODO but it might need to be a followup.